File: Java/Problem2.java
import java.util.HashMap;
import java.util.Map;

class LRUCache {

    private static class Node {
        int key;
        int value;
        Node prev;
        Node next;

    }

    private Map<Integer, Node> cache;
    private Node start;
    private Node end;
    private int size;

    public LRUCache(int capacity) {
        this.size = capacity;
        cache = new HashMap<>();
        start = new Node();
        end = new Node();
        this.start.next = this.end;
        this.end.prev = this.start;
    }

    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }

        Node node = cache.get(key);
        deleteNode(node);
        addFront(node);
        return node.value;

    }

    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            Node node = cache.get(key);
            node.value = value;
            deleteNode(node);
            addFront(node);
        } else {
            Node newnode = new Node();
            newnode.key = key;
            newnode.value = value;
            cache.put(key, newnode);
            addFront(newnode);
            if (cache.size() > size) {
                Node lru = end.prev;
                deleteNode(lru);
                cache.remove(lru.key);
            }
        }
    }

    private void deleteNode(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void addFront(Node node) {
        node.prev = this.start;
        node.next = this.start.next;
        this.start.next.prev = node;
        this.start.next = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: Python/Problem2.py
class Node:
    def __init__(self, key, val):
        self.key, self.val = key, val
        self.prev = self.next = None
        
class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache_map = {}

        self.least_recent, self.most_recent = Node(0,0), Node(0,0)
        self.least_recent.next, self.most_recent.prev = self.most_recent, self.least_recent

    def remove(self, node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self, node):
        prev, nxt = self.most_recent.prev, self.most_recent
        prev.next = nxt.prev = node
        node.next, node.prev = nxt, prev

    def get(self, key: int) -> int:
        if key in self.cache_map:
            #update the most recent use
            self.remove(self.cache_map[key])
            self.insert(self.cache_map[key])
            return self.cache_map[key].val
        else:
            return -1
    def put(self, key: int, value: int) -> None:
        if key in self.cache_map:
            self.remove(self.cache_map[key])
        self.cache_map[key] = Node(key, value)
        self.insert(self.cache_map[key])

        if len(self.cache_map) > self.capacity:
        #Evict LRU
            lru = self.least_recent.next
            self.remove(lru)
            del self.cache_map[lru.key]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)